compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -0,0 +1,1392 @@
+                        Type::Constant(0)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Lookup a given struct type by name.
+    fn lookup_struct_or_error(&mut self, path: Path) -> Option<Shared<StructType>> {
+        match self.lookup(path) {
+            Ok(struct_id) => Some(self.get_struct(struct_id)),
+            Err(error) => {
+                self.push_err(error);
+                None
+            }
+        }
+    }
+
+    /// Lookup a given trait by name/path.
+    fn lookup_trait_or_error(&mut self, path: Path) -> Option<&mut Trait> {
+        match self.lookup(path) {
+            Ok(trait_id) => Some(self.get_trait_mut(trait_id)),
+            Err(error) => {
+                self.push_err(error);
+                None
+            }
+        }
+    }
+
+    /// Looks up a given type by name.
+    /// This will also instantiate any struct types found.
+    pub(super) fn lookup_type_or_error(&mut self, path: Path) -> Option<Type> {
+        let ident = path.as_ident();
+        if ident.map_or(false, |i| i == SELF_TYPE_NAME) {
+            if let Some(typ) = &self.self_type {
+                return Some(typ.clone());
+            }
+        }
+
+        match self.lookup(path) {
+            Ok(struct_id) => {
+                let struct_type = self.get_struct(struct_id);
+                let generics = struct_type.borrow().instantiate(&mut self.interner);
+                Some(Type::Struct(struct_type, generics))
+            }
+            Err(error) => {
+                self.push_err(error);
+                None
+            }
+        }
+    }
+
+    fn lookup_type_alias(&mut self, path: Path) -> Option<Shared<TypeAlias>> {
+        self.lookup(path).ok().map(|id| self.interner.get_type_alias(id))
+    }
+
+    // this resolves Self::some_static_method, inside an impl block (where we don't have a concrete self_type)
+    fn resolve_trait_static_method_by_self(
+        &mut self,
+        path: &Path,
+    ) -> Option<(TraitMethodId, TraitConstraint, bool)> {
+        let trait_id = self.trait_id?;
+
+        if path.kind == PathKind::Plain && path.segments.len() == 2 {
+            let name = &path.segments[0].0.contents;
+            let method = &path.segments[1];
+
+            if name == SELF_TYPE_NAME {
+                let the_trait = self.interner.get_trait(trait_id);
+                let method = the_trait.find_method(method.0.contents.as_str())?;
+
+                let constraint = TraitConstraint {
+                    typ: self.self_type.clone()?,
+                    trait_generics: Type::from_generics(&the_trait.generics),
+                    trait_id,
+                };
+                return Some((method, constraint, false));
+            }
+        }
+        None
+    }
+
+    // this resolves TraitName::some_static_method
+    fn resolve_trait_static_method(
+        &mut self,
+        path: &Path,
+    ) -> Option<(TraitMethodId, TraitConstraint, bool)> {
```

Comment:
is this the same `bool` as in `resolve_trait_static_method_by_self`? if so, would it make sense to have a struct to keep track of these reused fields?

Reply:
Yep, it's the same. I'm avoiding refactorings in this code but I did add a comment explaining what it is used for.